### PR TITLE
W8 - D1 - HungTran - REST-API-with-Retrofit_and_Gson

### PR DESCRIPTION
**Pull request by at-hungtran2**
REST-API-with-Retrofit_and_Gson
#43

### CODE REVIEW

File: app/build.gradle [hunk 1]
```diff
@@ -68,4 +68,9 @@ dependencies {
     implementation 'com.android.support:design:27.1.1'
     apply from: "../ci-java.gradle"
     implementation 'com.android.support:design:27.1.1'
+    implementation 'com.squareup.retrofit2:retrofit:2.5.0'
+    implementation 'com.google.code.gson:gson:2.8.5'
+    implementation 'com.squareup.retrofit2:converter-gson:2.1.0'
```
Comment (taind89) [inline app/build.gradle]: please update to latest version

File: app/build.gradle [hunk 2]
```diff
@@ -68,4 +68,9 @@ dependencies {
...
+    implementation 'com.github.bumptech.glide:glide:4.7.1'
```
Comment (taind89) [inline app/build.gradle]: Please update to latest version
```suggestion
     implementation 'com.github.bumptech.glide:glide:4.8.0'
```

File: app/src/at-hungtran/java/asiantech/internship/summer/restapi/ImageAdapter.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+import com.bumptech.glide.Glide;
+
+import java.util.List;
+
+import asiantech.internship.summer.R;
+import asiantech.internship.summer.model.ImageItem;
+
+public class ImageAdapter extends RecyclerView.Adapter<ImageAdapter.ViewHolder> {
+    private List<ImageItem> mImageItems;
+    private Context mContext;
```
Comment (taind89) [inline app/src/at-hungtran/java/asiantech/internship/summer/restapi/ImageAdapter.java]: Context do not need to provide by constructor. You can get the `context` from itemView by `itemView.getContext()`

File: app/src/at-hungtran/java/asiantech/internship/summer/restapi/ImageAdapter.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
+        LayoutInflater layoutInflater = LayoutInflater.from(parent.getContext());
+        View itemView = layoutInflater.inflate(R.layout.item_image, parent, false);
+        return new ViewHolder(itemView);
+    }
+
+    @Override
+    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
+        ImageItem item = mImageItems.get(position);
+        Glide.with(mContext).load(item.getUrl()).into(holder.mImgPhoto);
```
Comment (taind89) [inline app/src/at-hungtran/java/asiantech/internship/summer/restapi/ImageAdapter.java]: Should has a function like `bindView` inside `ViewHolder` to handle binding data instead of handle binding data here

File: app/src/at-hungtran/java/asiantech/internship/summer/restapi/ImageAdapter.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+    public class ViewHolder extends RecyclerView.ViewHolder {
+        private ImageView mImgPhoto;
+        private TextView mTvContent;
+
+        public ViewHolder(View itemView) {
+            super(itemView);
+            mImgPhoto = itemView.findViewById(R.id.imgPhoto);
+            mTvContent = itemView.findViewById(R.id.tvContent);
+        }
```
Comment (taind89) [inline app/src/at-hungtran/java/asiantech/internship/summer/restapi/ImageAdapter.java]: Should has a `function` handle binding data here
Ex : 
```suggestion
         public void bindView(ImageItem item) {
            Glide.with(mContext).load(item.getUrl()).into(mImgPhoto);
            mTvContent.setText(item.getType());
        }
```

File: app/src/at-hungtran/java/asiantech/internship/summer/restapi/RestAPIActivity.java [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+    private static final int REQUEST_CAMERA = 1;
+    private static final int REQUEST_GALLERY = 2;
+
+    private static final int READ_PERMISSION_REQUEST_CAMERA = 1111;
+    private static final int WRITE_PERMISSION_REQUEST_CAMERA = 1112;
+    private static final int WRITE_PERMISSION_REQUEST_CALLERY = 11114;
+
+    private final CharSequence[] mChoiceOption = {"Camera", "Gallery"};
+    private APIImages mAPIImages;
+    private boolean mIsLoading;
```
Comment (taind89) [inline app/src/at-hungtran/java/asiantech/internship/summer/restapi/RestAPIActivity.java]: Please remove Resources unused

File: app/src/at-hungtran/java/asiantech/internship/summer/restapi/RestAPIActivity.java [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+    private int mCurrentPage;
+    private int mLastQueryImageNumber;
```
Comment (taind89) [inline app/src/at-hungtran/java/asiantech/internship/summer/restapi/RestAPIActivity.java]: Please remove Resources unused

File: app/src/at-hungtran/java/asiantech/internship/summer/restapi/RestAPIActivity.java [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+
+    private Uri getUriImageCamera(Bitmap photo) {
+        ByteArrayOutputStream bytes = new ByteArrayOutputStream();
+        photo.compress(Bitmap.CompressFormat.JPEG, 100, bytes);
+        String path = MediaStore.Images.Media.insertImage(getContentResolver(), photo, "Title", null);
+        return Uri.parse(path);
+    }
+
+    private void uploadImage(Uri imageUri) {
+        File file = new File(Objects.requireNonNull(getRealPathImage(imageUri)));
```
Comment (taind89) [inline app/src/at-hungtran/java/asiantech/internship/summer/restapi/RestAPIActivity.java]: Please show progressDialog for this task

File: app/src/at-hungtran/java/asiantech/internship/summer/restapi/RestAPIActivity.java [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+
+        RequestBody requestBody = RequestBody.create(MediaType.parse("multipart/form-data"), file);
+        MultipartBody.Part image = MultipartBody.Part.createFormData("imagedata", file.getName(), requestBody);
+
+        RequestBody token = RequestBody.create(MediaType.parse("text/plain"), APIImages.TOKEN);
+        mAPIImages.uploadImage(APIImages.UPLOAD_URL, token, image).enqueue(new Callback<ImageItem>() {
+            @Override
+            public void onResponse(@NonNull Call<ImageItem> call, @NonNull Response<ImageItem> response) {
+                if (response.isSuccessful()) {
+                    mImageAdapter.notifyItemInserted(0);
```
Comment (taind89) [inline app/src/at-hungtran/java/asiantech/internship/summer/restapi/RestAPIActivity.java]: Why you call `notifyItemInserted(0)` here. I don't see any item has been added to position 0 ?

File: app/src/at-hungtran/java/asiantech/internship/summer/restapi/RestAPIActivity.java [hunk 10]
```diff
@@ -0,0 +1,264 @@
...
+        mAPIImages.getImages(APIImages.TOKEN, mCurrentPage).enqueue(new Callback<List<ImageItem>>() {
+            @Override
+            public void onResponse(@NonNull Call<List<ImageItem>> call,
+                                   @NonNull Response<List<ImageItem>> response) {
+                if (response.isSuccessful()) {
+                    List<ImageItem> addImages = response.body();
+                    if (addImages != null) {
+                        mImages.addAll(addImages);
+                        mImageAdapter.notifyDataSetChanged();
+                        mLastQueryImageNumber = addImages.size();
```
Comment (taind89) [inline app/src/at-hungtran/java/asiantech/internship/summer/restapi/RestAPIActivity.java]: I see you update this `parameter` but I don't see where you use it ?

File: app/src/at-hungtran/java/asiantech/internship/summer/restapi/RestAPIActivity.java [hunk 11]
```diff
@@ -0,0 +1,264 @@
...
+                        mCurrentPage++;
+                        mIsLoading = false;
```
Comment (taind89) [inline app/src/at-hungtran/java/asiantech/internship/summer/restapi/RestAPIActivity.java]: Same comment for `mLastQueryImageNumber`

File: app/src/at-hungtran/res/layout/activity_rest_api.xml [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+            android:text="@string/loadImage" />
+
+        <Button
+            android:id="@+id/btnUploadImage"
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:text="@string/uploadImage" />
+    </LinearLayout>
+
+    <RelativeLayout
```
Comment (taind89) [inline app/src/at-hungtran/res/layout/activity_rest_api.xml]: I think this `Layout` is waste.I don't understand why you need this` Layout` here ?

File: app/src/at-hungtran/res/layout/activity_rest_api.xml [hunk 13]
```diff
@@ -0,0 +1,40 @@
+<LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
```
Comment (taind89) [inline app/src/at-hungtran/res/layout/activity_rest_api.xml]: Overall, this is a bad way to create this layout. Please arrange the nested layout as little as possible

File: app/src/at-hungtran/res/layout/item_image.xml [hunk 14]
```diff
@@ -0,0 +1,18 @@
...
+    <ImageView
+        android:id="@+id/imgPhoto"
+        android:layout_width="200dp"
+        android:layout_height="200dp" />
+    <TextView
+        android:id="@+id/tvContent"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_marginTop="20dp"/>
+</LinearLayout>
```
Comment (taind89) [inline app/src/at-hungtran/res/layout/item_image.xml]: Do not forget add break line here, thanks!

File: app/src/test/java/asiantech/internship/summer/ImageItemUnitTest.java [hunk 15]
```diff
@@ -5,11 +5,11 @@
 import static org.junit.Assert.*;
```
Comment (taind89) [inline app/src/test/java/asiantech/internship/summer/ImageItemUnitTest.java]: This file is not related to this PR. Why you change it ?

**Comment by taind89** [comment]
@at-hungtran2 
Please don't `push` the changes which not related to this PR

File: app/src/at-hungtran/java/asiantech/internship/summer/restapi/ImageAdapter.java [hunk 16]
```diff
@@ -0,0 +1,60 @@
+package asiantech.internship.summer.restapi;
+
+import android.content.Context;
```
Comment (at-dev-support-01) [inline app/src/at-hungtran/java/asiantech/internship/summer/restapi/ImageAdapter.java]: WARNING: [UnusedImports] Avoid unused imports such as 'android.content.Context'
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/imports.html#UnusedImports

File: app/src/at-hungtran/java/asiantech/internship/summer/restapi/RestAPIActivity.java [hunk 17]
```diff
@@ -239,6 +239,7 @@ private void uploadImage(Uri imageUri) {
             public void onResponse(@NonNull Call<ImageItem> call, @NonNull Response<ImageItem> response) {
                 if (response.isSuccessful()) {
                     mProgressDialog.dismiss();
+                    mImageAdapter.notifyItemInserted(0);
```
Comment (at-tainguyen) [inline app/src/at-hungtran/java/asiantech/internship/summer/restapi/RestAPIActivity.java]: Where is the source code which handles update data?
